### PR TITLE
Fix type of seen suggestions

### PR DESCRIPTION
**Pull request by nixel2007**
Given
* you set some completions in settings.cson in  'autocomplete' -> 'symbols' -> 'builtins' -> 'suggestions' section ([example](https://github.com/xDrivenDevelopment/atom-language-1c-bsl/blob/master/settings/language-1c-bsl.cson#L21))

---

When
* you type this function first time

Then
* you'll see a `b` mark and a description

When
* you type this function *not* first time

Then
* you'll see no mark or description

---

With this PR:

### CODE REVIEW

File: lib/symbol-store.js [hunk 1]
```diff
@@ -49,6 +49,21 @@ export default class SymbolStore {
     let firstLetter = prefix[0].toLowerCase()
     let symbolsByWord = new Map()
     let wordOccurrences = new Map()
+    let seen = new Map()
```
Comment (nixel2007) [inline lib/symbol-store.js]: Add a map, where we can check, if suggestion was already seen

File: lib/symbol-store.js [hunk 2]
```diff
@@ -49,6 +49,21 @@ export default class SymbolStore {
...
+    
+    let suggestions = []
+    for (let type of Object.keys(config)) {
+      let symbols = config[type].suggestions || EMPTY_ARRAY
+      for (let symbol of symbols) {
+        let {score} = this.scoreSymbol(prefix, symbol, cursorBufferRow, Number.MAX_VALUE)
+        if (score > 0) {
+          symbol.replacementPrefix = prefix
+          suggestions.push({symbol, score})
+          seen.set((symbol.text ? symbol.text : "") + (symbol.snippet ? symbol.snippet : ""), true)
```
Comment (nixel2007) [inline lib/symbol-store.js]: Fill the map

File: lib/symbol-store.js [hunk 3]
```diff
@@ -70,7 +85,9 @@ export default class SymbolStore {
               let type = symbol.matchingTypeForConfig(config)
               if (type != null) {
                 symbol = {text: symbol.text, type, replacementPrefix: prefix}
-                symbolsByWord.set(symbol.text, {symbol, score, localityScore})
+                if (!seen.get(symbol.text)) {
```
Comment (nixel2007) [inline lib/symbol-store.js]: Check if it was seen

**Comment by lee-dohm** [comment]
Couldn't this be solved by unchecking the "Enable Built-In Provider" setting?

screen shot 2016-02-25 at 12 51 18 pm

**Comment by benogle** [comment]
[The provider](https://github.com/atom/autocomplete-plus/wiki/Provider-API#defining-a-provider) should be able to suppress the builtin provider with the `excludeLowerPriority` property.

**Comment by nixel2007** [comment]
@lee-dohm it will disable the built-in provider :)
the problem is not that built-in provider interferes with language-provider. in that package there is no provider at all.
it's about wrong symbol detection in built-in provider.

**Comment by nixel2007** [comment]
ping

**Comment by nixel2007** [comment]
Any news on this?

**Comment by lee-dohm** [comment]
Thanks for the ping. We're pretty constrained lately because we have two core team developers out on parental leave :blush: I'll ask around and see if someone has time to take a look at this.

**Comment by nixel2007** [comment]
One more ping

**Comment by as-cii** [comment]
Thanks so much @nixel2007, this looks pretty good! 💯 Can we add a test that highlights the wrong behavior and ensures this pull request fixes it? After that, I think we can go ahead and 🚢 it!

**Comment by nixel2007** [comment]
@as-cii can i use my own grammar with built-in provider or you can point me to some standard grammar with built-ins, included in atom?

**Comment by as-cii** [comment]
> can point me to some standard grammar with built-ins, included in atom?

Totally, you should be able to use `language-javascript` as we did [elsewhere](https://github.com/atom/autocomplete-plus/blob/a3b73b735162632260de282f40308e262da74eef/spec/symbol-provider-spec.coffee#L33) in this package. 👍

**Comment by nixel2007** [comment]
@as-cii hm... i have some troubles with writing the spec.
The first - language-javascript don't contains any config.builtins. So i used [this grammar-package](https://atom.io/packages/language-1c-bsl) to test.

The second - what i did - checkout current master branch.
Add sample.bsl with:
```bsl
ValueFromFile();
```

Add this to some `describe` in `symbol-store-spec.coffee`
```coffeescript
    it "dont override description of built-in function", ->

      waitsForPromise ->
        Promise.all [
          atom.packages.activatePackage("language-1c-bsl")
          atom.workspace.open('sample.bsl').then (e) -> editor = e
        ]

      runs ->
        config =
          builtins:
            suggestions: [
              {
                'type': 'function'
                'rightLabel': 'global function'
                'text': 'ValueFromFile'
                'description': 'Test description.'
              }]

        editor.moveToBottom()
        occurrences = store.symbolsForConfig(config, null, 'val')

        expect(occurrences.length).toBe 1
        expect(occurrences[0].symbol.text).toBe 'ValueFromFile'
        expect(occurrences[0].symbol.description).toBe 'Test description.'
```

And... it passed =/ But definitely this bug still occurs. 
autocomplete

I will be grateful for any help with this =|

**Comment by nixel2007** [comment]
> The first - language-javascript don't contains any config.builtins. So i used this grammar-package to test.

My bad, i really don't need any package, cause i redefine the config. :)
I'll correct the test.

**Comment by as-cii** [comment]
I think the test needs to be changed to reuse the existing editor that is created [on these lines](https://github.com/atom/autocomplete-plus/blob/master/spec/symbol-store-spec.coffee#L5-L11). I assume that will make the test fail, which should help you debugging the problem more easily. 👍

**Comment by nixel2007** [comment]
@as-cii 
okey, i've made a test, that fails in master.

```coffeescript
    it "dont override description of built-in function", ->
      config =
        function:
          selectors: Selector.create('.function')
          typePriority: 1
        builtins:
          suggestions: [
            {
              'type': 'function'
              'rightLabel': 'global function'
              'text': 'ValueFromFile'
              'description': 'Test description.'
            }]

      editor.moveToBottom()
      editor.insertText('ValueFromFile()')

      occurrences = store.symbolsForConfig(config, [editor.getBuffer()], 'value')
      expect(occurrences.length).toBe 1
      expect(occurrences[0].symbol.text).toBe 'ValueFromFile'
      expect(occurrences[0].symbol.description).toBe 'Test description.'
      expect(occurrences[0].symbol.rightLabel).toBe 'global function'
```

But i have one more glitch:
autocomplete

I'll try to go deeper :)

**Comment by nixel2007** [comment]
@as-cii I've rebased my changes on latest master and last issue looks resolved.
Also I've added a test.

Please review.
Cheers!

File: spec/symbol-store-spec.coffee [hunk 4]
```diff
@@ -100,6 +100,29 @@ describe 'SymbolStore', ->
       expect(occurrences[0].symbol.text).toBe 'abc'
       expect(occurrences[0].symbol.type).toBe 'newtype'
 
+    it "dont override description of built-in function", ->
```
Comment (as-cii) [inline spec/symbol-store-spec.coffee]: What do you think about rewording this as `doesn't override built-in suggestions with the symbols found in the buffer`?

File: lib/symbol-store.js (hunk 2 above)
Comment (as-cii) [inline lib/symbol-store.js]: What do you think about changing this into a conditional?

```javascript
if (symbol.text) {
  seen.add(symbol.text)
} else if (symbol.snippet) {
  seen.add(symbol.snippet)
}
```

File: lib/symbol-store.js (hunk 1 above)
Comment (as-cii) [inline lib/symbol-store.js]: I think we should be able to use a `Set` here. Also, what about renaming the variable from `seen` to `builtinSymbolsByWord`?

**Comment by nixel2007** [comment]
@as-cii done! 👍

**Comment by as-cii** [comment]
Thanks @nixel2007! 💥

**Pull request merged**